enhancements/authentication/configuring-webhook-token-authenticators.md

```diff
@@ -0,0 +1,176 @@
+   secret in `openshift-kube-apiserver` namespace.
+3. If the webhook authenticator does not pass validation, the operator keeps the previous state.
+4. Redeploys the kube-apiserver static pods while specifying the new
+   mounted file path in the `apiServerArguments.authentication-token-webhook-config-file`
+
+#### Backend - Integrated OAuth Server Token Authentication
+
+1. cluster-authentication-operator observes `authentication.config/cluster` `type` field
+2. if set to "IntegratedOAuth", it creates an `openshift-config/webhook-authentication-integrated-oauth`
+  secret and adds it to the `webhookTokenAuthenticator` field of the `authentication.config/cluster`
+  resource
+  - the above secret contains details that point kube-apiserver to the oauth-apiserver for token reviews
+3. during an authentication flow, kube-apiserver sends `tokenreviews.authentication.k8s.io` object to the
+  oauth-apiserver for review
+4. oauth-apiserver sends a response accordingly to Kubernetes requirements, on success it adds the
+  `system:authenticated:oauth` group to the returned details about the groups of the authenticated user
+  - `system:authenticated:oauth` provides users with the ability to handle projects via the self-provisioner
+    cluster role
+  - similarly, if any integrator would like their users to have project privileges, they need to add this
+    specific group in their token review responses.
+
+### Common Gotchas When Providing Your Own Authentication Webhook
+
+1. `oauth-proxy` relies on the _/.well-known/oauth-authorization-server_ endpoint
+   and if the integrated oauth-server is turned off and the `spec.oauthMetadata` field
+   in `authentications.config/cluster` is not set properly, things like Grafana
+   and Prometheus won't be accessible
+2. ServiceAccounts may not be usable as OAuth clients anymore, which would result in
+  breaking authentication for 3rd party components that rely on this functionality
+  (e.g the Jenkins plugin integrating with OpenShift, but also Grafana and Prometheus
+  that also use service accounts as OAuth clients for oauth-proxy)
+
+It is the sole responsibility of the replacement component to make sure all of these work.
+
+## Design Details
+
+### Test Plan
+
+Since we will be making the cluster-authentication-operator use the `webhookTokenAuthenticator`
+(in combination with `oauthMetadata`, which already works) to configure authentication using
+the integrated OAuth server, we will just re-use all the tests that we have already.
+
+### Upgrade / Downgrade Strategy
+
+To keep authentication working even during upgrades, we'll keep our kube-apiserver
+authenticator patch code in place for 4.n version (the first version with the new wiring),
+but will disable it if any webhook authenticators are set. The same conditional OpenShift authentication
+should also appear in the 4.(n-1) version, which is the first version to contain the oauth-apiserver
```

Comment:
this I cannot follow. Why can't 4.(n-1) just have the patched authenticator, i.e. the webhook authenticator (if configured) is just ignored? Would be a good enough downgrade IMO.

Reply:
Agreed, even the oauth-apiserver appearing in an older version does not make sense, the whole sentence should just be removed.